### PR TITLE
Parse URL field in '.dzi' file

### PR DESCRIPTION
**Pull request by Linden6**
Normally when OpenSeadragon is called it is passed the location of a '.dzi' tile sources file located in remote file server.  This has the form:
```js
OpenSeadragon({
    id: 'img',
    prefixUrl: '/images/',
    tileSources: 'http://test.com/y/xy.dzi'
});
```

However, sometimes there is no '.dzi' file and the tile sources information is given directly within the page-source.  This then has the form:
```js
OpenSeadragon({
    id: 'img',
    prefixUrl: '/images/',
    tileSources: {
	Image: {
            type: 'image',
            xmlns: 'http://schemas.microsoft.com/deepzoom/2008',
	    Url: 'http://content.example.net/images/'
            Format:   'jpg', 
            Overlap:  '1', 
            TileSize: '254',
            Size: {
                Height: '4409',
                Width: '7793'
            }
	}
    }
});
```

With this second form a URL field is required so that OpenSeadragon knows where to locate the tiles.

This second form can be emulated by providing a link to a locally created '.dzi' file, such as:
```js
OpenSeadragon({
    id: 'img',
    prefixUrl: '/images/',
    tileSources: '/tmp/example.dzi'
});
```

However this local '.dzi' file needs to include a URL field, such as:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Image xmlns="http://schemas.microsoft.com/deepzoom/2009"
  Url="https://some/url/"
  Format="png"
  Overlap="1"
  TileSize="254"
  >
  <Size 
    Height="10000"
    Width="10000"
  />
</Image>
```

The proposed change allows the parsing of a 'Url' field, if present, in a '.dzi' file.  This field forms the base URL for tile requests and replaces the part up to and including '*_files' that would be present in a standard '.dzi' URL.

I would be grateful if you could tidy up the changes as I am sure that they can be made more elegant.

Many thanks!

### CODE REVIEW

File: src/dzi/dzi_file.rs [hunk 1]
```diff
@@ -14,6 +14,12 @@ pub struct DziFile {
     pub format: String,
     #[serde(rename = "Size", default)]
     pub sizes: Vec<Size>,
+    #[serde(rename = "Url", default = "no_url")]
+    pub base_url: String,
```
Comment (lovasoa) [inline src/dzi/dzi_file.rs:18]: This should probably be an `Option<String>` instead of a `String`.

**Review by lovasoa — CHANGES_REQUESTED** [review]

**Comment by lovasoa** [comment]
Hello, and thank you for the pull request ! I'm open to merging that, but it would need to be documented somewhere, since this new property is not standard. Ideally, we could directly parse the Openseadragon call from an HTML source, just like dezoomify does.

**Comment by lovasoa** [comment]
Do you have an example of a domain that has dezoomify tiles, but no `.dzi` file ?

**Comment by Linden6** [comment]
See this [page](https://openseadragon.github.io/examples/tilesource-dzi/) on OpenSeadragon's site - in particular the section entitled 'Inline Configuration for DZI':

**_'Note however the non-standard DZI property url which we must supply explicitly, since this is normally inferred from the path specified for the DZI XML/JSON/JSONP.'_**

**Comment by lovasoa** [comment]
> See this [page](https://openseadragon.github.io/examples/tilesource-dzi/) on OpenSeadragon's site - in particular the section entitled 'Inline Configuration for DZI'

All examples on this page link to images that actually have a hosted `.dzi` file. Do you have any example of a website that would not host the dzi file (and where this new feature would thus be useful) ?

**Comment by lovasoa** [comment]
> It is not always possible to parse OpenSeadragon calls from HTML source code as sometimes they are hidden away in javascript functions.

Being able to parse HTML would be a superset of this pull request. In cases where the HTML is not directly readable, it would still be possible to create a local HTML or javascript file.

**Comment by Linden6** [comment]
I do have a site in mind.  The images it hosts are generally out of copyright, but I would prefer not to unnecessarily ruffle any feathers by naming it.

It has taken a number of steps to prevent downloading of its images: not hosting dzi files, hiding OpenSeadragon calls in functions on external JavaScript files, and encrypting the values of the tile source fields.

**Comment by lovasoa** [comment]
I don't think this feature is worth merging if you are going to be its only user. On the contrary, we should add the name of your site to the README together with documentation of the feature, so that other than you can also download images from that site.

**Comment by Linden6** [comment]
If the site was based in a Western liberal democracy - where the legal consensus is that the photograph of a work of art in the public domain is unlikely to be subject to copyright - I'd be relaxed about sharing its name.

However, when as it's not, and belongs to a state institution in a country not renowned for its commitment to the rule of law, and in a country that you visit periodically, you might have reason to be a little more reticent.

**Comment by lovasoa** [comment]
Okay. Let's at least add information on the wiki about how to use this new feature, and link to it from the readme.

**Comment by lovasoa** [comment]
@Linden6 : I implemented the ability to parse openseadragon sources, for both dzi and iiif. 

I tested it on : 
  - https://openseadragon.github.io/examples/tilesource-iiif/
  - https://openseadragon.github.io/examples/tilesource-dzi/

If you don't find any issue, I'll merge it.

**Comment by Linden6** [comment]
These changes look great.  Thanks!

However, one important feature that needs implementing is to allow absolute - not just relative - URL's in the OpenSeadragon call.

If, for example, I create the following local 'image.html' file in the '/tmp' directory **with an absolute URL in the 'Url' field**:
```
<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Image Viewer</title>
    <style>
      #viewer {
	  min-height: 300px;
	  height: 95vh;
      }
    </style>
    <script src="https://openseadragon.github.io/openseadragon/openseadragon.min.js"></script>
  </head>
  <body>

    <div id="viewer"></div>

    <script>
      OpenSeadragon({
	  id:              "viewer",
	  prefixUrl:       "https://openseadragon.github.io/openseadragon/images/",
	  tileSources: {
              Image: {
		  xmlns:    "http://schemas.microsoft.com/deepzoom/2008",
		  Url:      "https://openseadragon.github.io/example-images/highsmith/highsmith_files/",
		  Format:   "jpg", 
		  Overlap:  "2", 
		  TileSize: "256",
		  Size: {
                      Height: "9221",
                      Width:  "7026"
		  }
              }
	  }
      });
    </script>
  </body>
</html>
```

I can link to it in my browser (`file:///tmp/viewer.html`) and the viewer with its image appear fine.

However, at the moment if I call 'dezoomify-rs /tmp/image.html image.jpg' then the following error messages appear:

> Unable to download tile '/tmp/imagehtmlhttps://openseadragon.github.io/example-images/highsmith/highsmith_files/11/0_0.jpg'
> Unable to download tile '/tmp/imagehtmlhttps://openseadragon.github.io/example-images/highsmith/highsmith_files/11/0_1.jpg'
> 

Allowing absolute URL's is important as - more often than not - where tile source information is given via inline configuration code it cannot be simply parsed like these OpenSeadragon pages (https://openseadragon.github.io/examples/tilesource-dzi/ &  https://openseadragon.github.io/examples/tilesource-iiif/).  It will be necessary for the user to carrying out their own investigations and enter the tile source information that they have gleaned into a local html file with an absolute 'Url' field that can then be given as the first argument to dezoomify-rs.

Thanks again

**Comment by Linden6** [comment]
I've noticed a bug when trying to convert from Deep Zoom images from to IIIF.  For example,
`dezoomify-rs https://openseadragon.github.io/examples/tilesource-dzi/ image.iiif`
results in an image that is mostly black, and a multitude of warning messages appear, starting:

> [2020-09-12T14:38:37Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=512 y=512 at zoom level 1 and position x=1536 y=512 was not fully covered by source tiles. It misses 260608 pixels.
> [2020-09-12T14:38:37Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=512 y=512 at zoom level 1 and position x=4096 y=3072 was not fully covered by source tiles. It misses 260096 pixels.
> [2020-09-12T14:38:37Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=512 y=512 at zoom level 1 and position x=1024 y=7168 was not fully covered by source tiles. It misses 260608 pixels.
> [2020-09-12T14:38:38Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=512 y=512 at zoom level 1 and position x=3584 y=1024 was not fully covered by source tiles. It misses 261120 pixels.
> [2020-09-12T14:38:38Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=512 y=512 at zoom level 1 and position x=5632 y=7168 was not fully covered by source tiles. It misses 260608 pixels.

A couple of associated images:
a
(fully zoomed out)

b
(zoomed in)

Creating normal image files, for example
`dezoomify-rs https://openseadragon.github.io/examples/tilesource-dzi/ image.jpg`, is fine.

The problem only appears to arise with DZI -> IIIF conversation.  For example
`dezoomify-rs https://openseadragon.github.io/examples/tilesource-iiif/ image.iiif` is fine.

Thanks

**Comment by lovasoa** [comment]
> However, one important feature that needs implementing is to allow absolute - not just relative - URL's in the OpenSeadragon call.

I fixed that. Resolving http URLs now works even when the source file is on the local filesystem.

**Comment by lovasoa** [comment]
> I've just noticed a bug when trying to convert from OpenSeadragon inline configured Deep Zoom images

The current implementation of the retiler does not handle overlapping tiles, which some Deep Zoom images use. When it encounters pixels that it has already written, it does not recognize it, and starts a new tile, that will contain only the overlapping pixels.

**Comment by lovasoa** [comment]
> The current implementation of the retiler does not handle overlapping tiles

I fixed that in 503da49

**Comment by Linden6** [comment]
That's great - many many thanks!

**Review by lovasoa — APPROVED** [review]
OK, I'll merge when the tests finish, then. And I'll make a new release next week if no one finds bugs.

**Comment by Linden6** [comment]
I've just run `dezoomify-rs file.html image.iiif` a couple of times at different zoom levels.

It now works perfectly with an absolute URL in a local http file.

And pixels now cover **nearly** all the source tiles.  However when calling the first Deep Zoom source from 
https://openseadragon.github.io/examples/tilesource-dzi/ at zoom level 2 (1757 x  2306 pixels) via:
```<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <title>Image Viewer</title>
    <style>
      #viewer {
	  min-height: 300px;
	  height: 95vh;
      }
    </style>
    <script src="https://openseadragon.github.io/openseadragon/openseadragon.min.js"></script>
  </head>
  <body>

    <div id="viewer"></div>

    <script>
      OpenSeadragon({
	  id:              "viewer",
	  prefixUrl:       "https://openseadragon.github.io/openseadragon/images/",
	  tileSources: {
              Image: {
		  xmlns:    "http://schemas.microsoft.com/deepzoom/2008",
		  Url:      "https://openseadragon.github.io/example-images/highsmith/highsmith_files/",
		  Format:   "jpg", 
		  Overlap:  "2", 
		  TileSize: "256",
		  Size: {
                      Height: "9221",
                      Width:  "7026"
		  }
              }
	  }
      });
    </script>
  </body>
</html>
```
The following two warnings were received:

> [2020-09-12T17:53:24Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=1757 y=2048 at zoom level 4 and position x=0 y=0 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T17:53:25Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=1757 y=258 at zoom level 4 and position x=0 y=2048 was not fully covered by source tiles. It misses 504 pixels.

With the two most zoomed out tiles:
./0,0,1757,2048/440,512/0/default.jpg  - having a 1px right black edge
./0,2048,1757,258/440,65/0/default.jpg - having a 1px right and bottom black edge

And when calling the same command and choosing zoom level 0, the following warning messages were received:

> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=2048 at zoom level 4 and position x=6144 y=6144 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2048 y=1029 at zoom level 4 and position x=4096 y=8192 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=2048 at zoom level 4 and position x=6144 y=4096 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2048 y=1029 at zoom level 4 and position x=2048 y=8192 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:13Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=1029 at zoom level 4 and position x=6144 y=8192 was not fully covered by source tiles. It misses 478 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2048 y=1029 at zoom level 4 and position x=0 y=8192 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=2048 at zoom level 4 and position x=6144 y=2048 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=882 y=2048 at zoom level 4 and position x=6144 y=0 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2930 y=4096 at zoom level 8 and position x=4096 y=0 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=4096 y=1029 at zoom level 8 and position x=0 y=8192 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:14Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2930 y=1029 at zoom level 8 and position x=4096 y=8192 was not fully covered by source tiles. It misses 495 pixels.
> [2020-09-12T18:41:15Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=2930 y=4096 at zoom level 8 and position x=4096 y=4096 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:15Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=7026 y=8192 at zoom level 16 and position x=0 y=0 was not fully covered by source tiles. It misses 512 pixels.
> [2020-09-12T18:41:15Z WARN  dezoomify_rs::encoder::retiler] The target tile of size x=7026 y=1029 at zoom level 16 and position x=0 y=8192 was not fully covered by source tiles. It misses 504 pixels.

Again these tiles are edge tiles with edges along the right or bottom of the image.

The overall dimensions of the tiles appear correct, so I'm not sure it's a rounding error this time.

Thanks!

**Pull request merged**